From 5b6936af3f4c0d0f3af55372c426d91c0a5b5a35 Mon Sep 17 00:00:00 2001
From: "Banka Sai Krishna(Temp)" <quic_banksaik@quicinc.com>
Date: Mon, 15 Jul 2024 15:54:21 +0530
Subject: [PATCH] hostapd: Fix TPE ext duplicate IEs

When all subbands have the same PSD values, TPE EXTs are being filled
for all subbands. Address this issue by filling TPE EXTs for all
subbands only if the PSD values differ. If not, fill with a single
PSD value.

Patch-depdendency: r21-001-hostapd-TPE-11AX-changes-and-support-for-11BE-TPE.patch

Patch-work: none

Signed-off-by: Banka Sai Krishna <quic_banksaik@quicinc.com>
---
 src/ap/ieee802_11.c | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7749,15 +7749,6 @@ static int get_psd_values(struct hostapd
 			is_different_psd = 1;
 	}
 
-	if (!is_different_psd && !ru_punct_bitmap) {
-		*tx_pwr_count = 0;
-	} else {
-		*tx_pwr_count = num_psd_values_to_psd_count(non_11be_chan_count);
-		if (*tx_pwr_count == 0) {
-			wpa_printf(MSG_ERROR, "Invalid channel count:%d", non_11be_chan_count);
-			return -1;
-		}
-	}
 #ifdef CONFIG_IEEE80211BE
 	/* For 11be the TPE extension parameter added if the bw is 320MHZ or if
 	 * any channel is punctured in 320MHZ/160MHZ/80MHZ
@@ -7782,8 +7773,20 @@ static int get_psd_values(struct hostapd
 		}
 		tx_pwr_ext_array++;
 		*tx_pwr_ext_count += 1;
+		if (!is_different_psd && (start_chan_psd != chan_psd))
+			is_different_psd = 1;
 	}
 #endif
+	if (!is_different_psd && !ru_punct_bitmap) {
+		*tx_pwr_count = 0;
+		*tx_pwr_ext_count = 0;
+	} else {
+		*tx_pwr_count = num_psd_values_to_psd_count(non_11be_chan_count);
+		if (*tx_pwr_count == 0) {
+			wpa_printf(MSG_ERROR, "Invalid channel count:%d", non_11be_chan_count);
+			return -1;
+		}
+	}
 	return 0;
 }
 
